File: gui/ovencon.py
#! /usr/bin/python

'''
Oven controller GUI - works in conjunction with external oven control hardware.

Copyright (c) 2011, Daniel Strother < http://danstrother.com/ >
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
  - Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  - Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
  - The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS OR IMPLIED
WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
'''

# depends on python-qt4, python-qwt5-qt4, python-serial
from PyQt4 import QtGui, QtCore
import PyQt4.Qwt5 as Qwt
import math
import sys
import serial
import time

class OvenMsg():
    """Class representing a single status message sent from the oven control hardware."""

    def __init__(self):
        """Sets sane default message contents."""
        self.state      = 'idle'
        self.time       = 0.0
        self.target     = 0.0
        self.sense_t    = 0.0
        self.sense_b    = 0.0
        self.cmd        = 0.0
        self.cmd_t      = 0.0
        self.cmd_b      = 0.0

    def parse(self,msg):
        """Parses message contents from a comma-separated string.
        
        On microcontroller, message is generated with the C code:
        sprintf_P(tx_msg,PSTR("%s,%u,%d,%d,%d,%u,%u,%u\\n"),
            state_names[state],
            time,
            target,
            temp_t,
            temp_b,
            cmd,
            cmd_t,
            cmd_b);
        
        This parses that."""

        m               = msg.split(',')
        if(len(m) != 8):
            return 0

        self.state      = m.pop(0)
        self.time       = int(m.pop(0))*0.25
        self.target     = int(m.pop(0))*0.25
        self.sense_t    = int(m.pop(0))*0.25
        self.sense_b    = int(m.pop(0))*0.25
        self.cmd        = int(m.pop(0))/255.0
        self.cmd_t      = int(m.pop(0))/255.0
        self.cmd_b      = int(m.pop(0))/255.0

        return 1


class OvenCommThread(QtCore.QThread):
    """Thread class responsible for asynchronously receiving messages from oven controller on behalf of OvenComm instance."""

    def __init__(self,parent=None):
        """Constructor; requires an OvenComm parent."""
        super(OvenCommThread,self).__init__(parent)
        self.running = True
        self.p = parent

    def __del__(self):
        """Tells thread to stop, then waits (a little while) for thread to terminate"""
        self.stop()
        self.wait()

    def run(self):
        """Triggers newMessage signal in parent OvenComm instance whenever a message is received."""
        while(self.running):
            msg = OvenMsg()
            if(msg.parse(self.p.s.readline()) and self.running):
                self.p.trigger_newMessage(msg)

    def stop(self):
        """Tells comm thread to stop running (eventually)."""
        self.running = False


class OvenComm(QtCore.QObject):
    """Class that encapsulates all serial communication with oven controller hardware."""

    newMessage = QtCore.pyqtSignal('PyQt_PyObject')     # PyQt_PyObject required for passing non-C++ types through Qt signals

    newSenseT = QtCore.pyqtSignal(float)
    newSenseB = QtCore.pyqtSignal(float)

    def __init__(self,parent=None,port='/dev/ttyUSB000'):
        """Opens specified serial port and starts comm thread."""

        super(OvenComm,self).__init__(parent)
        
        self.thread = None
        self.s = None

        self.s = serial.Serial(port=port,timeout=0.7)

        # clear any stale data that may have been buffered prior to program start
        time.sleep(0.5)
        self.s.flushInput()
        self.s.readline()

        self.thread = OvenCommThread(self)
        self.thread.start()

        self.v_cmd_t = 0
        self.v_cmd_b = 0

    def __del__(self):
        """Terminates comm thread and closes serial port."""

        self.newSenseT.disconnect()
        self.newSenseB.disconnect()
        self.newMessage.disconnect()

        if(self.thread):
            self.thread.stop()
            self.thread.wait()  # don't close serial port until after listening thread has stopped
        if(self.s):
            self.s.close()

    def trigger_newMessage(self,msg):
        """Callback for triggering Qt signals on receipt of new message by OvenCommThread."""

        self.newSenseT.emit(msg.sense_t)
        self.newSenseB.emit(msg.sense_b)
        self.newMessage.emit(msg)

    def go(self):
        """Callback for Go button - resets controller and starts reflow operation."""
        self.s.write("reset\ngo\n")

    def reset(self):
        """Callback for Reset button - just resets controller (stops ongoing reflow operation)."""
        self.s.write("reset\n")

    def pause(self):
        """Callback for Pause button - puts controller into pause state (profile time does not advance)."""
        self.s.write("pause\n")

    def resume(self):
        """Callback for Resume button - resumes controller after pause (profile time resumes advancing)."""
        self.s.write("resume\n")

    def manual(self,m):
        """Callback for Manual check-box - when enabled, controller's PID loop is bypassed."""
        if(m):
            self.s.write("manual: 1\n")
        else:
            self.s.write("manual: 0\n")

    def cmd_t(self,t):
        """Callback for Command-Top slider - controls power to top heating element."""
        self.v_cmd_t = int(t*255.0/100.0)
        self.cmd()

    def cmd_b(self,b):
        """Callback for Command-Bottom slider - controls power to bottom heating element."""
        self.v_cmd_b = int(b*255.0/100.0)
        self.cmd()

    def cmd(self):
        """Sends updated manual power commands to controller (after clamping to safe ranges)."""
        if(self.v_cmd_t < 0):
            self.v_cmd_t = 0
        if(self.v_cmd_t > 255):
            self.v_cmd_t = 255
        if(self.v_cmd_b < 0):
            self.v_cmd_b = 0
        if(self.v_cmd_b > 255):
            self.v_cmd_b = 255
        print >> self.s, "cmd: %d, %d" % (self.v_cmd_t,self.v_cmd_b)

    def target(self,t):
        """Callback for Target slider - controls target temperature when in Idle state."""
        t = int(t*4)
        if(t < 0):
            t = 0
        if(t > 4095):
            t = 4095
        print >> self.s, "target: %d" % (t)


class OvenLogger():
    """Class for logging oven status messages to CSV files."""

    def __init__(self,comm):
        """Connects newMessage signal handler to OvenComm instance."""
        self.comm = comm
        self.prevstate = 'idle'
        self.f = None
        self.time_offset = 0.0
        self.comm.newMessage.connect(self.log_message)

    def __del__(self):
        """Closes open log file."""
        if(self.f):
            self.f.close()

    def start_new_file(self):
        """Starts a new time-stamped log file and inserts column headers."""
        if(self.f):
            self.f.close()
        filename = time.strftime('ovenlog_%Y%m%d%H%M%S.csv')
        self.f = open(filename,'a')
        print >> self.f, "state,time,target,sense_t,sense_b,cmd,cmd_t,cmd_b"

    def log_message(self,msg):
        """Callback for newMessage signals - writes messages to log file.

        Opens a new log file when transitioning from a non-idle state to idle."""

        if(not self.f or (msg.state == 'idle' and self.prevstate != 'idle')):
            # when transitioning to idle state (or if log file not already open),
            # open a new log file
            self.start_new_file()

            # time_offset is used to make all log files start at time 0
            # (controller time always increments, and never resets)
            self.time_offset = msg.time

        self.prevstate = msg.state
        print >> self.f, "%s,%f,%f,%f,%f,%f,%f,%f" % (msg.state,msg.time-self.time_offset,msg.target,msg.sense_t,msg.sense_b,msg.cmd,msg.cmd_t,msg.cmd_b)


class OvenPlot(Qwt.QwtPlot):
    """Common base-class for plotting oven data."""

    def __init__(self,comm,parent=None):
        """Connents newMessage handler to OvenComm instance and sets up common plot format."""
        super(OvenPlot,self).__init__(parent)

        self.comm = comm
        
        self.setCanvasBackground(QtCore.Qt.white)

        # plot grid
        grid = Qwt.QwtPlotGrid()
        pen = QtGui.QPen(QtCore.Qt.DotLine)
        pen.setColor(QtCore.Qt.black)
        pen.setWidth(0)
        grid.setPen(pen)
        grid.attach(self)

        # plot legend and x-axis (y-axis handled in derived classes)
        self.insertLegend(Qwt.QwtLegend(), Qwt.QwtPlot.BottomLegend)
        self.setAxisTitle(Qwt.QwtPlot.xBottom, "Time (seconds)")

        self.times = []
        self.time_offset = 0.0
        self.max_idle = (120*4)
        self.prevstate = 'idle'

        self.reset_plot()
        
        comm.newMessage.connect(self.newMessage_handler)

    def newMessage_handler(self,msg):
        """Callback for newMessage signals - adds new data to plot."""

        if(msg.state == 'idle' and self.prevstate != 'idle'):
            # when transitioning to idle state, start with a clean plot
            self.time_offset = msg.time
            self.reset_plot()

        self.prevstate = msg.state
        self.update_plot(msg)

    def update_plot(self,msg):
        """Adds new data to plot, and prunes old data."""

        self.times.append(msg.time-self.time_offset)

        if(msg.state == 'idle'):
            # when idle, newest entry is always at time 0
            # (so transition to non-idle state begins at time 0)
            diff = msg.time - self.time_offset
            self.time_offset = msg.time
            self.times = [t - diff for t in self.times]

        if(msg.state == 'idle' and len(self.times) > self.max_idle):
            # when idle, limit visible window of time to max_idle entries
            self.times.pop(0)

    def reset_plot(self):
        """Resets plot to empty (no data) state."""
        self.times = []


class OvenTempPlot(OvenPlot):
    """Class for plotting temperatures in oven."""

    def __init__(self,comm,parent=None):
        """Sets up OvenTempPlot-specific formatting."""

        super(OvenTempPlot,self).__init__(comm,parent)

        self.setTitle("Temperatures")
        self.setAxisTitle(Qwt.QwtPlot.yLeft, "Temperature (degrees celsius)")

        self.c_target = Qwt.QwtPlotCurve("Target")
        self.c_sense_t = Qwt.QwtPlotCurve("Top")
        self.c_sense_b = Qwt.QwtPlotCurve("Bottom")

        self.c_target.attach(self)
        self.c_sense_t.attach(self)
        self.c_sense_b.attach(self)
        
        pen = QtGui.QPen()
        pen.setColor(QtCore.Qt.black)
        pen.setWidth(4)
       
        self.c_target.setPen(pen)
        self.c_sense_t.setPen(QtGui.QPen(QtCore.Qt.red))
        self.c_sense_b.setPen(QtGui.QPen(QtCore.Qt.blue))

    def update_plot(self,msg):
        """Adds new data to plot, and prunes old data."""

        super(OvenTempPlot,self).update_plot(msg)

        self.target.append(msg.target)
        self.sense_t.append(msg.sense_t)
        self.sense_b.append(msg.sense_b)

        if(msg.state == 'idle' and len(self.target) > self.max_idle):
            # when idle, limit visible window of time to max_idle entries
            self.target.pop(0)
            self.sense_t.pop(0)
            self.sense_b.pop(0)

        self.c_target.setData(self.times,self.target)
        self.c_sense_t.setData(self.times,self.sense_t)
        self.c_sense_b.setData(self.times,self.sense_b)

        # actually update the plot
        self.replot()

    def reset_plot(self):
        """Resets plot to empty (no data) state."""

        super(OvenTempPlot,self).reset_plot()

        self.target = []
        self.sense_t = []
        self.sense_b = []


class OvenCommandPlot(OvenPlot):
    """Class for plotting power commands to oven."""

    def __init__(self,comm,parent=None):
        """Sets up OvenCommandPlot-specific formatting."""

        super(OvenCommandPlot,self).__init__(comm,parent)

        self.setTitle("Commands")
        self.setAxisTitle(Qwt.QwtPlot.yLeft, "Command")

        self.c_cmd = Qwt.QwtPlotCurve("Command")
        self.c_cmd_t = Qwt.QwtPlotCurve("Top")
        self.c_cmd_b = Qwt.QwtPlotCurve("Bottom")

        self.c_cmd.attach(self)
        self.c_cmd_t.attach(self)
        self.c_cmd_b.attach(self)
        
        pen = QtGui.QPen()
        pen.setColor(QtCore.Qt.black)
        pen.setWidth(4)
       
        self.c_cmd.setPen(pen)
        self.c_cmd_t.setPen(QtGui.QPen(QtCore.Qt.red))
        self.c_cmd_b.setPen(QtGui.QPen(QtCore.Qt.blue))

    def update_plot(self,msg):
        """Adds new data to plot, and prunes old data."""

        super(OvenCommandPlot,self).update_plot(msg)

        self.cmd.append(msg.cmd*100.0)
        self.cmd_t.append(msg.cmd_t*100.0)
        self.cmd_b.append(msg.cmd_b*100.0)

        if(msg.state == 'idle' and len(self.cmd) > self.max_idle):
            # when idle, limit visible window of time to max_idle entries
            self.cmd.pop(0)
            self.cmd_t.pop(0)
            self.cmd_b.pop(0)

        self.c_cmd.setData(self.times,self.cmd)
        self.c_cmd_t.setData(self.times,self.cmd_t)
        self.c_cmd_b.setData(self.times,self.cmd_b)

        # actually update the plot
        self.replot()

    def reset_plot(self):
        """Resets plot to empty (no data) state."""

        super(OvenCommandPlot,self).reset_plot()

        self.cmd = []
        self.cmd_t = []
        self.cmd_b = []


class OvenPlots(QtGui.QWidget):
    """Widget containing all GUI plots for oven controller."""

    def __init__(self,comm,parent=None):
        """Creates plots and assigns layouts."""

        super(OvenPlots,self).__init__(parent)

        self.temps = OvenTempPlot(comm,parent=self)
        self.commands = OvenCommandPlot(comm,parent=self)

        l = QtGui.QVBoxLayout()
        l.addWidget(self.temps)
        l.addWidget(self.commands)

        self.setLayout(l)


class OvenControls(QtGui.QWidget):
    """Widget containing all GUI controls for oven controller."""

    def __init__(self,comm,parent=None):
        """Creates controls and assigns callbacks/layouts."""

        super(OvenControls,self).__init__(parent)

        # slider for idle-state target temperature
        self.target_slider = QtGui.QSlider(QtCore.Qt.Vertical,self)
        self.target_slider.setMinimum(0)
        self.target_slider.setMaximum(250)
        self.target_slider.valueChanged.connect(comm.target)

        # slider for manual-mode top element power
        self.cmdt_slider = QtGui.QSlider(QtCore.Qt.Vertical,self)
        self.cmdt_slider.setMinimum(0)
        self.cmdt_slider.setMaximum(100)
        self.cmdt_slider.valueChanged.connect(comm.cmd_t)

        # slider for manual-mode bottom element power
        self.cmdb_slider = QtGui.QSlider(QtCore.Qt.Vertical,self)
        self.cmdb_slider.setMinimum(0)
        self.cmdb_slider.setMaximum(100)
        self.cmdb_slider.valueChanged.connect(comm.cmd_b)

        # sliders adjacent to each other
        sliders = QtGui.QHBoxLayout()
        sliders.addWidget(self.target_slider)
        sliders.addWidget(self.cmdt_slider)
        sliders.addWidget(self.cmdb_slider)

        # LCDs indicating sensed temperatures
        self.senset_lcd = QtGui.QLCDNumber(self)
        self.senseb_lcd = QtGui.QLCDNumber(self)
        self.senset_lcd.setSegmentStyle(QtGui.QLCDNumber.Flat)
        self.senseb_lcd.setSegmentStyle(QtGui.QLCDNumber.Flat)
        comm.newSenseT.connect(self.senset_lcd.display)
        comm.newSenseB.connect(self.senseb_lcd.display)

        self.manual_checkbox = QtGui.QCheckBox("Manual",self)
        self.manual_checkbox.stateChanged.connect(comm.manual)

        self.reset_button = QtGui.QPushButton("Reset",self)
        self.reset_button.clicked.connect(comm.reset)

        self.go_button = QtGui.QPushButton("Go",self)
        self.go_button.clicked.connect(comm.go)

        self.pause_button = QtGui.QPushButton("Pause",self)
        self.pause_button.clicked.connect(comm.pause)

        self.resume_button = QtGui.QPushButton("Resume",self)
        self.resume_button.clicked.connect(comm.resume)

        # overall column layout
        l = QtGui.QVBoxLayout()
        l.addWidget(self.senset_lcd)
        l.addWidget(self.senseb_lcd)
        l.addLayout(sliders)
        l.addWidget(self.manual_checkbox)
        l.addWidget(self.reset_button)
        l.addWidget(self.go_button)
        l.addWidget(self.pause_button)
        l.addWidget(self.resume_button)

        self.setLayout(l)


class OvenMain(QtGui.QWidget):
    """Widget containing all oven controller GUI elements."""

    def __init__(self,comm,parent=None):
        """Creates Plots and Controls, and assigns layout."""

        super(OvenMain,self).__init__(parent)
        
        self.plots      = OvenPlots(comm,parent=self)
        self.controls   = OvenControls(comm,parent=self)

        l = QtGui.QHBoxLayout()
        l.addWidget(self.plots)
        l.addWidget(self.controls)
        self.setLayout(l)


class OvenCon(QtGui.QMainWindow):
    """Main window for oven controller GUI."""

    def __init__(self,port):
        """Sets up main windows and starts application execution."""

        super(OvenCon,self).__init__()
        
        self.resize(1000,600)
        self.setWindowTitle('Reflow Controller')

        # connect to controller hardware
        self.comm       = OvenComm(parent=self,port=port)

        # log controller status to disk
        self.logger     = OvenLogger(self.comm)

        # create GUI
        self.main       = OvenMain(self.comm,parent=self)

        # associate GUI with window
        self.setCentralWidget(self.main)

if __name__ == '__main__':
    # start GUI application when invoked stand-alone
    app = QtGui.QApplication(sys.argv)
    port = '/dev/ttyUSB000';
    if(len(sys.argv)>1 and sys.argv[1]):
        # get serial port from command line
        port = sys.argv[1]
    try:
        qb = OvenCon(port)
    except serial.SerialException as se:
        print "failed to open serial port - \"%s\"" % (se)
        sys.exit(1)
    else:
        qb.show()
        sys.exit(app.exec_())

File: avr/ovencon.c
/**
 * Copyright (c) 2011, Daniel Strother < http://danstrother.com/ >
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *   - Redistributions of source code must retain the above copyright notice,
 *     this list of conditions and the following disclaimer.
 *   - Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *   - The name of the author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "ovencon.h"
#include <util/delay.h>
#include <stdint.h>

#include <string.h>
#include <stdio.h>

#include "usb_serial.h"

#include "oven_ssr.h"
#include "oven_timing.h"
#include "oven_pid.h"
#include "oven_profile.h"
#include "max6675.h"


// commands/config to controller from serial

volatile uint8_t mode_fake_out;
volatile uint8_t mode_fake_in;
volatile uint8_t mode_manual;

// TODO: there is a risk that some of these 16-bit values could be caught
// in a half-updated state, if the timer interrupt causes oven_update_4hz
// to be executed in the middle of process_message.
// Since these values are generally only used for testing, and not during
// actual reflow operations, this may not be a major concern.
volatile int16_t fake_temp_t;
volatile int16_t fake_temp_b;
volatile uint8_t manual_cmd_t;
volatile uint8_t manual_cmd_b;
volatile int16_t manual_target;



extern volatile uint8_t k_p;
extern volatile uint8_t k_i;
extern volatile uint8_t k_d;



#define CMD_RESET   1
#define CMD_GO      2
#define CMD_PAUSE   3
#define CMD_RESUME  4

// TODO: currently, only one comm_cmd can be processed per oven_update_4hz
// invocation, so multiple commands received within a ~0.25s window may be lost
volatile uint8_t comm_cmd;


// controller state

#define ST_FAULT    0
#define ST_IDLE     1
#define ST_RUN      2
#define ST_DONE     3
#define ST_PAUSE    4

const char *state_names[5] = { "fault","idle","run","done","pause" };

uint8_t state = ST_FAULT;

int16_t target;
uint16_t time;


char tx_msg[255];
volatile uint8_t tx_len = 0;

void fault(void)
{
// TODO: need to mask faults briefly at power-up, otherwise this will
// prematurely trip due to bogus thermocouple data
//    state = ST_FAULT;
//    ssr_fault();

    tx_len=sprintf_P(tx_msg,PSTR("FAULT\n"));
    if (!usb_configured()) return;
    usb_serial_write((void*)tx_msg,tx_len);
    tx_len = 0; // clear the length, so the control loop knows it can generate a new message
}

void thermocouple_fault(int16_t result)
{
    tx_len = sprintf_P(tx_msg,PSTR("TFAULT: %d\n"),
	            result);
    if (!usb_configured()) return;
     usb_serial_write((void*)tx_msg,tx_len);
     tx_len = 0; // clear the length, so the control loop knows it can generate a new message
}

void debugmsg(PGM_P  pmsg){
#ifdef DEBUG
	if (!usb_configured()) return;

    if (tx_len>0)
        usb_serial_write((void*)tx_msg,tx_len);
    tx_len = sprintf_P(tx_msg,pmsg);
    usb_serial_write((void*)tx_msg,tx_len);
#endif
    //tx_msg[0]=0;
   // tx_len = 0; // clear the length, so the control loop knows it can generate a new message
   //  _delay_ms(100);

}

void oven_output(uint8_t top, uint8_t bot)
{
    if( !mode_fake_out )
    {
        ssr_set(top,bot);
    }
    else
    {
        ssr_set(0,0);
    }
}

void oven_input(int16_t *top, int16_t *bot)
{
    if( !mode_fake_in )
    {
    	*top = max6675_read(0);

#ifndef BOTTOM_THERM
    	*bot = *top;
#else
        *bot = max6675_read(1);
#endif
    }
    else
    {
        *top = fake_temp_t;
        *bot = fake_temp_b;
    }
}

void oven_setup(void)
{
    state = ST_IDLE;

    mode_fake_out   = 0;
    mode_fake_in    = 0;
    mode_manual     = 0;

    manual_cmd_t    = 0;
    manual_cmd_b    = 0;
    manual_target   = 0;
    fake_temp_t     = 0;
    fake_temp_b     = 0;

    k_p   = DEFAULT_K_P;
    k_i   = DEFAULT_K_I;
    k_d   = DEFAULT_K_D;

    comm_cmd        = 0;

    target          = 0;
    time            = 0;
    tx_len          = 0;

    pid_reset();
    profile_reset();
    ssr_setup();
    max6675_setup();

    max6675_start();
    timing_setup();
}

void oven_update_120hz(void)
{
    ssr_update();
}


void oven_update_4hz(void)
{
    int16_t temp_t,temp_b;
    uint8_t cmd,cmd_t,cmd_b;
   
    oven_input(&temp_t,&temp_b);

    if(comm_cmd != 0)
    {
        switch(comm_cmd)
        {
            case CMD_RESET:
                profile_reset();
                pid_reset();
                manual_target   = 0;
                manual_cmd_t    = 0;
                manual_cmd_b    = 0;
                state           = ST_IDLE;
                break;
            case CMD_GO:
                if(state == ST_IDLE) {
                    state           = ST_RUN;
                }
                break;
            case CMD_PAUSE:
                if(state == ST_RUN) {
                    state           = ST_PAUSE;
                }
                break;
            case CMD_RESUME:
                if(state == ST_PAUSE) {
                    state           = ST_RUN;
                }
                break;
            default:
                fault();
        }

        comm_cmd = 0;
    }

    switch(state)
    {
        case ST_IDLE:
            target = manual_target;
            break;
        case ST_RUN:
            if(profile_update(&target))
                state = ST_DONE;
            break;
        case ST_PAUSE:
            // hold target
            break;
        case ST_DONE:
            target = 0;
            break;
        default:
            fault();
    }

    // manual target will always track target (so there aren't any surprises
    // when enabling manual mode)
    manual_target = target;
    
    cmd = pid_update(temp_t,target);

    if( state == ST_IDLE && mode_manual )
    {
        // full manual control from serial port
        cmd_t = manual_cmd_t;
        cmd_b = manual_cmd_b;
    }
    else
    {
        // manual power commands zeroed when not in full-manual mode (again: no surprises)
        manual_cmd_t = 0;
        manual_cmd_b = 0;

        // 25/75 top/bottom split (until bottom saturates)
        cmd_t = cmd >> 2;
        cmd_b = cmd - cmd_t;

        if(cmd_b >= 127) {
            cmd_t += (cmd_b - 127);
            cmd_b = 255;
        } else {
            cmd_b <<= 1;
        }

        if(cmd_t >= 127) {
            cmd_t = 255;
        } else {
            cmd_t <<= 1;
        }
    }

    oven_output(cmd_t,cmd_b);

    // produce status update message
    // (provided previous update has already been sent)
    if(!tx_len)
    {
        // expensive.. but we're only running at 4 Hz, so we have cycles to burn
        tx_len = sprintf_P(tx_msg,PSTR("%s,%u,%d,%d,%d,%u,%u,%u\n"),
            state_names[state],
            time,
            target,
            temp_t,
            temp_b,
            cmd,
            cmd_t,
            cmd_b);
    }

    time++;

}

char rx_msg[255];
uint8_t rx_cnt;

void process_message(const char *msg)
{
    // this is a ridiculously expensive function to invoke - a more efficient
    // command parser could be implemented, or a binary protocol established -
    // but, we're not expecting a lot of command traffic in this application,
    // and all of the timing-critical routines are handled by interrupts, so
    // there isn't a lot of downside to this expensive-but-easy implementation


    cli(); // temporarily disable interrupts to prevent any potential write errors

    if(sscanf_P(msg,PSTR("temp: %d, %d"),&fake_temp_t,&fake_temp_b) || \
       sscanf_P(msg,PSTR("cmd: %hhu, %hhu"),&manual_cmd_t,&manual_cmd_b) || \
       sscanf_P(msg,PSTR("target: %d"),&manual_target) || \
       sscanf_P(msg,PSTR("fake_out: %hhu"),&mode_fake_out) || \
       sscanf_P(msg,PSTR("fake_in: %hhu"),&mode_fake_in) || \
       sscanf_P(msg,PSTR("manual: %hhu"),&mode_manual) ||
       sscanf_P(msg,PSTR("pid: %d, %d, %d"),&k_p, &k_i, &k_d)) {
        ;
    } else if(strcmp_P(msg,PSTR("reset")) == 0) {
        comm_cmd = CMD_RESET;
    } else if(strcmp_P(msg,PSTR("go")) == 0) {
        comm_cmd = CMD_GO;
    } else if(strcmp_P(msg,PSTR("pause")) == 0) {
        comm_cmd = CMD_PAUSE;
    } else if(strcmp_P(msg,PSTR("resume")) == 0) {
        comm_cmd = CMD_RESUME;
    }

    sei();
}

// program entry point
int main(void)
{
    int16_t ret;
    char c;

    // no prescaler
    CLKPR = 0x80;
    CLKPR = 0;

    // hold CS high until init
    DDRB|=_BV(0);
    PORTB|=_BV(0);


    // initialize
    oven_setup();


    // sleep.  Makes the USB less cranky
    _delay_ms(1000);

    // wait for usb to initialize
    usb_init();
    while (!usb_configured());


    // wait an arbitrary bit for the host to complete its side of the init
    _delay_ms(1000);
    //timing_setup(); // timing setup last, since it enables timer interrupts (which invoke the update functions)

    // clear any stale packets
    usb_serial_flush_input();
    rx_cnt = 0;



    // run forever
    while(1)
    {
        // if the control loop has generated a status update message,
        // send it out over USB to the host
        if(tx_len)
        {
            usb_serial_write((void*)tx_msg,tx_len);
            tx_len = 0; // clear the length, so the control loop knows it can generate a new message
        }

        // receive individual characters from the host
        while( (ret = usb_serial_getchar()) != -1)
        {
            // all commands are terminated with a new-line
            c = ret;
            if(c == '\n') {
                // only process commands that haven't overflowed the buffer
                if(rx_cnt > 0 && rx_cnt < 255) {
                    rx_msg[rx_cnt] = '\0';
                    process_message(rx_msg);
                }
                rx_cnt = 0;
            } else {
                // buffer received characters
                rx_msg[rx_cnt] = c;
                if(rx_cnt != 255) rx_cnt++;
            }
        }
    }
}

